### PR TITLE
MESOS: add node/slave-update controller to controller-manager

### PR DESCRIPTION
**Pull request by sttts**
The status of a node is updated by the kubelet usually. If the kubelet doesn't do that in time, the node controller will set the Ready condition to false.

In the Mesos case the kubelet is only running for those node where a pod is running. Hence, we have to take over the status update from the scheduler for those nodes. In this PR we add a StatusUpdater to the scheduler doing this. The Mesos cloud provider is extended to return nodes without a running kubelet (it has this information from the state.json anyway). Instead of "KubeletReady" the new controller posts "SlaveReady".

Fixes https://github.com/mesosphere/kubernetes-mesos/issues/498

### CODE REVIEW

**Comment by jdef** [comment]
Some initial thoughts/questions:
(a) picking an interval between that of the kubelet and node controller is clever - and smells fragile. upon my first pass through the code it's not clear to me whether this will avoid some kind of condition "flap" resulting from a strange interaction of these various timers.

(b) i'm not super excited about seeing this control process being introduced into the scheduler, but i'm even less excited about the additional cloud provider dependency (.. and another implicit dep on mesos/state.json); we should be steering away from additional queries polling of master, not adding more.

(c) the scheduler already has access to the collection of slaves in the cluster via incoming offers. why not just use a node/slave list built from that instead of polling mesos master via cloudprovider? i realize that offers aren't guaranteed to come in from all slaves if resources are not available, but then we're not going to be launching things on those types of slaves anyway. if we go this route, maybe we could change the condition from `SlaveReady` to `SlaveOfferingResources`? it may take time for the condition status of slaves to converge across the cluster but I think I can live with that. slaves that aren't generating offers, but already have k8s tasks running will be running a kubelet that's updating node condition, so nothing to worry about there.

**Comment by jdef** [comment]
related to https://github.com/mesosphere/kubernetes-mesos/issues/498?

**Comment by sttts** [comment]
(a) depends on the right choice of timers I guess. The monitoring and kubelet updating interval is 5 sec at the moment, compared to 40 sec grace period. Possibly we can also improve the situation with the knowledge which node runs an executor.

(b) in fact we can do the same from the controller manager. The choice to do it from the scheduler was due to the idea in (c) to use the offer.

(c) as written above, I was thinking about the same approach using the offers. This works fine with 5 nodes, but not with 100 or 1000. We will not get offers for a large number of them. Moreover, we won't get offers for nodes which are saturated with tasks (if mesos master is clever enough not to send us offers for nodes without left memory or cpu).

Another reason why we "have" to use the state.json slave list is that we have to stop heartbeating node which are not existing anymore. Of course, we could use yet another huge grace period using the offers and delete node resources if there hasn't been an offer for some time. But this really sounds fragile.

To sum up:

I think we should move this logic into our own MesosSlaveController to avoid this new state.json dependency in the scheduler. Moreover, what we build is a controller and belongs into the Controller Manager just by that reason.

In the state.json we also have the right information to know where kubelets are running. If we expose this information somehow we can avoid your concerns regarding fragility due to timeout/heartbeat races. Though I am not convinced yet we need such an advanced logic at all.

**Comment by jdef** [comment]
> I think we should move this logic into our own MesosSlaveController to avoid this new state.json dependency in the scheduler. Moreover, what we build is a controller and belongs into the Controller Manager just by that reason.

Is this still the plan going forward?

**Comment by sttts** [comment]
@jdef compare the last commits. It's implemented now. What is left is a more clever timeout handling.

**Comment by sttts** [comment]
Rebase and gofmt fixes.

File: contrib/mesos/pkg/node/statusupdater.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+	newNodeReadyCondition := api.NodeCondition{
+		Type:              api.NodeReady,
+		Status:            api.ConditionTrue,
+		Reason:            "SlaveReady",
+		Message:           "slave is up",
+		LastHeartbeatTime: currentTime,
+	}
+
+	updated := false
+	for i := range n.Status.Conditions {
```
Comment (jdef) [inline contrib/mesos/pkg/node/statusupdater.go]: why not use `getCondition()` here?

File: pkg/cloudprovider/providers/mesos/client.go [hunk 2]
```diff
@@ -55,7 +55,7 @@ type slaveNode struct {
 
 type mesosState struct {
 	clusterName string
-	nodes       []*slaveNode
+	nodes       map[string]*slaveNode
```
Comment (jdef) [inline pkg/cloudprovider/providers/mesos/client.go]: please doc the key value (looks like node hostname?)

File: contrib/mesos/pkg/node/statusupdater.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+		if err := u.tryUpdateStatus(n); err != nil {
+			log.Errorf("Error updating node status, will retry: %v", err)
+		} else {
+			return nil
+		}
+	}
+	return fmt.Errorf("Update node status exceeds retry count")
+}
+
+func (u *StatusUpdater) tryUpdateStatus(n *api.Node) error {
```
Comment (jdef) [inline contrib/mesos/pkg/node/statusupdater.go]: this looks sufficiently complex that a unit test would be nice.

File: contrib/mesos/pkg/node/statusupdater.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+				continue
+			}
+
+			err := u.updateStatus(&nodes[i])
+			if err != nil {
+				log.Errorf("Error updating node status: %v", err)
+			}
+		}
+	}
+	go runtime.Until(monitor, app.DefaultNodeMonitorPeriod, terminate)
```
Comment (jdef) [inline contrib/mesos/pkg/node/statusupdater.go]: instead of DefaultNodeMonitorPeriod, should this be some derivative of relistPeriod? (or, perhaps, should be set to the value of relistPeriod?)

**Comment by jdef** [comment]
still have concerns about timing. perhaps they can be mitigated by the addition of some nice docs as to how we expect the various timers to interact?

File: contrib/mesos/pkg/node/statusupdater.go (hunk 1 above)
Comment (sttts) [inline contrib/mesos/pkg/node/statusupdater.go]: I guess because it's also updating the condition, not only getting it.

File: contrib/mesos/pkg/node/statusupdater.go (hunk 3 above)
Comment (sttts) [inline contrib/mesos/pkg/node/statusupdater.go]: done

File: pkg/cloudprovider/providers/mesos/client.go (hunk 2 above)
Comment (sttts) [inline pkg/cloudprovider/providers/mesos/client.go]: done

File: contrib/mesos/pkg/node/statusupdater.go (hunk 4 above)
Comment (sttts) [inline contrib/mesos/pkg/node/statusupdater.go]: This is the kubelet status update frequency now.

**Comment by sttts** [comment]
@jdef I replaced this tricky timing now with a mechanism to detect nodes without kubelets now.

**Comment by sttts** [comment]
@k8s-bot test this please

**Comment by sttts** [comment]
Shippable timing out during the compile phase. wtf.

**Comment by sttts** [comment]
@jdef PTAL

File: contrib/mesos/pkg/node/statusupdater.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+		}
+	}
+	go runtime.Until(monitor, kubelet.DefaultNodeStatusUpdateFrequency, terminate)
+
+	return nil
+}
+
+func (u *StatusUpdater) updateStatus(n *api.Node) error {
+	for i := 0; i < nodeStatusUpdateRetry; i++ {
+		if err := u.tryUpdateStatus(n); err != nil {
```
Comment (jdef) [inline contrib/mesos/pkg/node/statusupdater.go]: probably don't want to retry if it's an error due to "conflict" or "invalid" spec. see https://github.com/kubernetes/kubernetes/blob/master/pkg/api/errors/errors.go `IsConflict`, `IsInvalid`

File: contrib/mesos/pkg/node/statusupdater.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+	for i := range n.Status.Conditions {
+		c := &n.Status.Conditions[i]
+		if c.Type == api.NodeReady {
+			if c.Status == newNodeReadyCondition.Status {
+				newNodeReadyCondition.LastTransitionTime = c.LastTransitionTime
+			} else {
+				newNodeReadyCondition.LastTransitionTime = currentTime
+			}
+			n.Status.Conditions[i] = newNodeReadyCondition
+			found = true
```
Comment (jdef) [inline contrib/mesos/pkg/node/statusupdater.go]: should this `break` once `found` is set to `true`?

File: contrib/mesos/pkg/node/statusupdater.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+// nodeWithUpdatedStatus clones the given node and updates the NodeReady condition.
+// The updated node is return and a boolean indicating whether the node was changed
+// at all.
+func (u *StatusUpdater) nodeWithUpdatedStatus(n *api.Node) (*api.Node, bool, error) {
+	readyCondition := getCondition(&n.Status, api.NodeReady)
+	currentTime := unversioned.Now()
+
+	// avoid flapping by waiting at least twice the kubetlet update frequency
```
Comment (jdef) [inline contrib/mesos/pkg/node/statusupdater.go]: so the kubelet is updating the LastHeartbeatTime periodically, and thinking here is that if we miss two heartbeats then we should proceed?

File: contrib/mesos/pkg/node/statusupdater_test.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+					Reason:             r,
+					Message:            "some message we don't care about here",
+					LastTransitionTime: unversioned.Time{time.Now().Add(-time.Minute)},
+					LastHeartbeatTime:  unversioned.Time{time.Now().Add(d)},
+				}},
+			},
+		}
+	}
+
+	assert.True(t, kubelet.DefaultNodeStatusUpdateFrequency*3 < app.DefaultNodeMonitorGracePeriod) // sanity check for defaults
```
Comment (jdef) [inline contrib/mesos/pkg/node/statusupdater_test.go]: :+1:

File: pkg/cloudprovider/providers/mesos/client.go [hunk 9]
```diff
@@ -50,12 +50,13 @@ type mesosClient struct {
 
 type slaveNode struct {
 	hostname  string
+	kubelet   bool // true iff a kubelet is running
```
Comment (jdef) [inline pkg/cloudprovider/providers/mesos/client.go]: nit: naming. consider `hasKubelet` or `kubeletRunning`

File: pkg/cloudprovider/providers/mesos/client.go [hunk 10]
```diff
@@ -50,12 +50,13 @@ type mesosClient struct {
...
 	resources *api.NodeResources
 }
 
 type mesosState struct {
 	clusterName string
-	nodes       []*slaveNode
+	nodes       map[string]*slaveNode // by hostname
```
Comment (jdef) [inline pkg/cloudprovider/providers/mesos/client.go]: :+1:

File: pkg/cloudprovider/providers/mesos/client.go [hunk 11]
```diff
@@ -225,12 +226,36 @@ func parseMesosState(blob []byte) (*mesosState, error) {
...
-	nodes := []*slaveNode{}
+
+	executorSlaveIds := map[string]struct{}{}
+	for _, f := range state.Frameworks {
+		for _, e := range f.Executors {
+			// Note that this simple comparison breaks when we support more than one
+			// k8s instance in a cluster. At the moment this is not possible for
+			// a number of reasons.
+			// TODO(sttts): find way to detect executors of this k8s instance
+			if e.Name == "Kubelet-Executor" {
```
Comment (jdef) [inline pkg/cloudprovider/providers/mesos/client.go]: isn't `Kubelet-Executor` a const somewhere? if not then perhaps it should be

File: contrib/mesos/pkg/node/statusupdater.go (hunk 6 above)
Comment (sttts) [inline contrib/mesos/pkg/node/statusupdater.go]: Break would make sense. I loosely follow `Kubelet.setNodeStatus` here.

File: contrib/mesos/pkg/node/statusupdater.go (hunk 7 above)
Comment (sttts) [inline contrib/mesos/pkg/node/statusupdater.go]: Yes. We poll the master's state.json. But due to caching this might be out of date. We give the kubelet two intervals to update the heartbeat. Only when the heartbeat does not appear twice and state.json tells us that the node is gone, we start heartbeating here.

File: pkg/cloudprovider/providers/mesos/client.go (hunk 9 above)
Comment (sttts) [inline pkg/cloudprovider/providers/mesos/client.go]: right.

File: pkg/cloudprovider/providers/mesos/client.go (hunk 11 above)
Comment (sttts) [inline pkg/cloudprovider/providers/mesos/client.go]: Here in the cloud manager we cannot reference anything in the contrib directory, unfortunately.

File: contrib/mesos/pkg/node/statusupdater.go (hunk 7 above)
Comment (jdef) [inline contrib/mesos/pkg/node/statusupdater.go]: ok sounds good. can we add that level of detail to the comments here?

**Comment by jdef** [comment]
minor comments. overall looks good

**Comment by jdef** [comment]
lgtm

**Comment by gmarek** [comment]
Sorry for not taking it earlier - I'm on vacation till the end of the week.

**Comment by sttts** [comment]
@gmarek enjoy your vacations.

@mikedanese can you take a look meanwhile? It's only a little constant refactoring.

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 12]
```diff
@@ -173,6 +173,12 @@ type VolumeConfigFlags struct {
 	PersistentVolumeRecyclerIncrementTimeoutHostPath    int
 }
 
+const (
+	DefaultNodeMonitorGracePeriod = 40 * time.Second
+	DefaultNodeStartupGracePeriod = 60 * time.Second
+	DefaultNodeMonitorPeriod      = 5 * time.Second
```
Comment (mikedanese) [inline cmd/kube-controller-manager/app/controllermanager.go]: Can you put these with all the other defaults (in NewCMServer, then reference them with s.NodeMonitorGracePeriod)?

Comment (sttts) [inline cmd/kube-controller-manager/app/controllermanager.go]: done.

**Comment by sttts** [comment]
@k8s-bot test this

**Comment by sttts** [comment]
@k8s-bot test this

**Comment by sttts** [comment]
@k8s-bot test this

**Comment by gmarek** [comment]
@k8s-bot test this please

Failure looks like flake.

**Comment by gmarek** [comment]
LGTM if tests will finally pass...

**Comment by sttts** [comment]
@gmarek thanks for the review.

e2e tests seem to be broken. 2 days ago we got https://github.com/kubernetes/kubernetes/issues/17256. Now the tests do not even start: https://goo.gl/Zps2Th.

**Comment by gmarek** [comment]
It passed, so we're good.

**Comment by gmarek** [comment]
LGTM, as @jdef already gave his.

**Comment by sttts** [comment]
Squashed.
